### PR TITLE
Switch most of CI to GitHub Actions

### PR DESCRIPTION
**Pull request by encukou**
This includes PR #38, where Travis CI is taking a really long time to run all the builds.

### CODE REVIEW

**Comment by encukou** [comment]
@hroncok, would you have time for a quick sanity check?

**Comment by encukou** [comment]
The Travis CI build failed on Python 3.7, which is no longer compatible with `c90` – the error comes from a CPython header.
The new GH action removes `py37-c90`.

**Comment by hroncok** [comment]
> This includes PR #38

GitHub presents the thing in a confusing manner. That PR appears to be merged but the commits are still displayed here. Anyway, reviewing the CI related commit only here.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+on:
+  push:
+    branches:
+    - master
+    - ci-test
+  pull_request:
+    branches:
+    - master
```
Comment (hroncok) [inline .github/workflows/main.yml:8]: Why limiting the branches here?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      uses: fedora-python/tox-github-action@master
+      with:
+        tox_env: ${{ matrix.tox_env }}
+        dnf_install: /usr/bin/make
+    strategy:
+      matrix:
+        tox_env: [
+            py27-c90, py35-c90, py36-c90,
+            py27-c99, py35-c99, py36-c99, py37-c99, py38-c99, py39-c99, py310-c99,
+            py27-c90, py35-cpp, py36-cpp, py37-cpp, py38-cpp, py39-cpp, py310-cpp,
```
Comment (hroncok) [inline .github/workflows/main.yml:28]:
```suggestion
            py27-cpp, py35-cpp, py36-cpp, py37-cpp, py38-cpp, py39-cpp, py310-cpp,
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  tox_test:
+    name: Tox test
+    steps:
+    - uses: actions/checkout@v2
+    - name: Run Tox tests
+      id: test
+      uses: fedora-python/tox-github-action@master
+      with:
+        tox_env: ${{ matrix.tox_env }}
+        dnf_install: /usr/bin/make
```
Comment (hroncok) [inline .github/workflows/main.yml:22]: Explicit is better than implicit, but you are wasting time by loading all the repodata only to see:

```
Package make-1:4.3-2.fc33.x86_64 is already installed.
```

Suggestion: Let's document and ensure make is installed in https://github.com/fedora-python/fedora-python-tox

**Review by hroncok — COMMENTED** [review]
> The Travis CI build failed on Python 3.7, which is no longer compatible with c90 – the error comes from a CPython header.
The new GH action removes py37-c90.

No commit message explicitly mentions this (fine by me if fine by you).

----

Python 3.4 is in tox config but not on the CI. It was enabled on Travis. Does GitHub Action lack 3.4, or is this an oversight?

----

`tox --parallel=auto` blows up due to race conditions in `test/build/lib.linux-x86_64-3.9` and similar directories. Not a big deal, but I thought you should know if you haven't considered it.

----

I've noticed that the cpp jobs mixes gcc and g++:

```
gcc -pthread -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -D_GNU_SOURCE -fPIC -fwrapv -Werror -Wall -Wno-error=strict-aliasing -fPIC -I../include -I/github/workspace/.tox/py36-cpp/include -I/usr/include/python3.6m -c test_py3c.cpp -o build/temp.linux-x86_64-3.6/test_py3c.o -Werror -Wall
creating build/lib.linux-x86_64-3.6
g++ -pthread -shared -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -g -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -g -Werror -Wall -Wno-error=strict-aliasing build/temp.linux-x86_64-3.6/test_py3c.o -L/usr/lib64 -lpython3.6m -o build/lib.linux-x86_64-3.6/test_py3c.cpython-36m-x86_64-linux-gnu.so
```

This does not seem to be caused by this PR, but is it intended?

File: .github/workflows/main.yml (hunk 3 above)
Comment (encukou) [reply .github/workflows/main.yml:22]: OK, I opened https://github.com/fedora-python/fedora-python-tox/issues/43

**Comment by encukou** [comment]
Thanks for the review!

The first `gcc` actually uses C++ because of the extension of `test_py3c.cpp` (which is a symlink to `test_py3c.c`, made just for this).

**Comment by encukou** [comment]
Alas, Python 3.4 is no longer in Fedora.

**Comment by hroncok** [comment]
Right, OMG.

Dne čt 7. 1. 2021 19:03 uživatel Petr Viktorin <notifications@github.com>
napsal:

**Comment by encukou** [comment]
OK! Looks like we can use Travis for 3.4 and an ancient container image for Python 3.3.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    container: python:3.3
+    steps:
+    - uses: actions/checkout@v2
+    - name: Run tests
+      env:
+        CFLAGS: "-Werror -Wall -Wno-error=strict-aliasing"
+      run: |
+        CFLAGS="$CFLAGS -std=c90" make test-python
+        CFLAGS="$CFLAGS -std=c99" make test-python
+        make test-python-cpp
```
Comment (hroncok) [inline .github/workflows/main.yml:41]: This is the cpp job? It's a tad hard to realize.

**Review by hroncok — COMMENTED** [review]
I wonder whether it won't make more sense to do the 3.4 job on GH actions as well?

File: .github/workflows/main.yml (hunk 4 above)
Comment (encukou) [reply .github/workflows/main.yml:41]: What do you suggest? It says `-cpp` at the end.

**Comment by encukou** [comment]
> I wonder whether it won't make more sense to do the 3.4 job on GH actions as well?

Gerrit Huizenga convinced me to keep Travis CI, or at least a bit of it. See [this discussion](https://github.com/encukou/py3c/pull/33#issuecomment-756211294) for the details.

**Comment by encukou** [comment]
Thank you for the review and all the suggestions! But I don't want to drag this out longer. This PR now makes things better than they were, so I'll merge.
Discussion can continue, of course :)

**Pull request merged**